### PR TITLE
feat(composables): add 100% coverage

### PR DESCRIPTION
**Pull request by mdanilowicz**

### CODE REVIEW

File: packages/composables/src/useProductPrice.ts [hunk 1]
```diff
@@ -71,37 +71,49 @@ export function useProductPrice(
...
-  const displayFrom: ComputedRef<boolean> = computed(
-    () =>
-      (product.value?.calculatedPrices?.length ?? 0) > 1 ||
-      !!(_displayParent.value && displayFromVariants.value),
-  );
+  const displayFrom: ComputedRef<boolean> = computed(() => {
+    // Change me mind:
+    //  `!!(_displayParent.value && displayFromVariants.value)` will never return true
+    // - _displayParent.value expects `product.value?.parentId === null`
+    // - displayFromVariants.value expects `!!product.value?.parentId`
```
Comment (patzick) [inline packages/composables/src/useProductPrice.ts:84]: yes, looks like a dead case, let's remove it

File: packages/composables/src/useProductPrice.ts [hunk 2]
```diff
@@ -71,37 +71,49 @@ export function useProductPrice(
...
-        );
-        return (
-          lowest || (_cheapest.value as unknown as Schemas["CalculatedPrice"])
-        );
+        return product.value?.calculatedPrices?.reduce((previous, current) => {
+          return current.unitPrice < previous.unitPrice ? current : previous;
+        });
+
+        // Change my mind:
+        // There is no possibility to return _cheapest.value because of the getProductTierPrices that is iterating over calculatedPrices
```
Comment (patzick) [inline packages/composables/src/useProductPrice.ts:113]: If somehow in the `calculatedPrices` the `unitPrice` would be `0` then in theory it is possible to return `_cheapest`. If that is the case anywhere? I'm not sure. I'd leave this logic (without casting to unknown and CalculatedPrice) and create a follow-up issue for the discovery of that case

Comment (mkucmus) [reply packages/composables/src/useProductPrice.ts:113]: now, when the types reflect the actual shape of the data from API we can trust it, so removing redundant code can be done. I've checked the preview and compared the variants, from prices and so on. The prices are displayed correctly - in comparison to `main` branch which reflects the storefront behavior.

File: packages/composables/src/useProduct.ts [hunk 3]
```diff
@@ -31,10 +30,7 @@ export function useProduct(
   configurator?: Ref<Schemas["PropertyGroup"][]> | Schemas["PropertyGroup"][],
 ): UseProductReturn {
   const _product = useContext("product", { context: product });
-  if (!_product.value) {
-    // TODO link docs with composables context usage
-    throw new ContextError("Product");
-  }
```
Comment (patzick) [inline packages/composables/src/useProduct.ts:37]: why is this removed?
If product is not provided with param or context it would cause runtime errors

File: packages/composables/src/useProductAssociations.spec.ts [hunk 4]
```diff
@@ -88,4 +90,18 @@ describe("useProductAssociations", () => {
 
     expect(vm.productAssociations).toStrictEqual(mockedCrossSelling);
   });
+
+  it("init without product", () => {
```
Comment (patzick) [inline packages/composables/src/useProductAssociations.spec.ts:94]:
```suggestion
  it("init without product should throw an error", () => {
```

File: packages/composables/src/useProductAssociations.spec.ts [hunk 5]
```diff
@@ -88,4 +90,18 @@ describe("useProductAssociations", () => {
...
+    expect(() =>
+      useSetup(() =>
+        useProductAssociations(
+          // Api case
+          ref(null) as unknown as ComputedRef<Schemas["Product"]>,
```
Comment (patzick) [inline packages/composables/src/useProductAssociations.spec.ts:99]:
```suggestion
          // @ts-expect-error if API returns null we want to be prepared for it
          ref(null),
```

File: packages/composables/src/useProductPrice.test.ts [hunk 6]
```diff
@@ -63,4 +62,65 @@ describe("useProductPrice", () => {
...
+            },
+          ) as unknown as Schemas["Product"],
+        ),
+      ),
+    );
+    expect(vm.displayFrom).toBe(true);
+    expect(vm.isListPrice).toBe(false);
+  });
+
+  it("isListPrice - 3", () => {
```
Comment (patzick) [inline packages/composables/src/useProductPrice.test.ts:91]: 2345678765456
better documented please :)

File: packages/composables/vitest.config.ts [hunk 7]
```diff
@@ -7,14 +7,20 @@ export default defineConfig({
...
-      // "100": true, // TODO: our goal ;)
       thresholds: {
-        statements: 97,
-        branches: 90,
-        functions: 88.4,
-        lines: 97,
+        statements: 100,
+        branches: 100,
+        functions: 100,
+        lines: 100,
```
Comment (patzick) [inline packages/composables/vitest.config.ts:14]: we can replace it with that
```suggestion
        "100": true,
```

File: packages/composables/vitest.config.ts [hunk 8]
```diff
@@ -7,14 +7,20 @@ export default defineConfig({
...
       },
-      exclude: ["**/devtools/**"],
+      exclude: [
+        "**/devtools/**",
+        "**/types/**",
+        "**/src/*.test.ts",
+        "**/src/*.spec.ts",
+        // After refactoring should be removed
```
Comment (patzick) [inline packages/composables/vitest.config.ts:21]: link to GH issue so it's known when resolved when someone will see this TODO in future

File: packages/composables/src/_test/useSetup.ts [hunk 9]
```diff
@@ -8,6 +8,7 @@ type Injections = {
   apiClient: {
     invoke: ReturnType<typeof vi.fn>;
   };
+  [key: string]: unknown;
```
Comment (patzick) [inline packages/composables/src/_test/useSetup.ts:11]: why do we need this generic injections? Would be better to know exactly what we're dealing with

**Review by patzick — CHANGES_REQUESTED** [review]
Great work 🙌  I reviewed the code and left some comments. Mainly to make test names more descriptive. People should not need to read the test code tested code to understand what the test is doing. This way, we also have proper documentation for the tested code.

While applying suggestions, please make sure that:
- test names are properly documenting features/behaviours of the code
- mocks are cleared between tests, so there is no impact between singular tests
- fill the PR description appropriately so when someone looks at this PR in the future description summarizes what's changed here

File: packages/composables/src/useNotifications.ts [hunk 10]
```diff
@@ -93,7 +93,7 @@ export function useNotifications(): UseNotificationsReturn {
     const messageId = geterateId();
     _notifications.value.push({
       id: messageId,
-      type: options.type || "info",
+      type: options.type as NotificationType,
```
Comment (patzick) [inline packages/composables/src/useNotifications.ts:96]: this casting is incorrect. `type` is optional so we had `info` as default type, now it would send `undefined as NotificationType`

Comment (patzick) [reply packages/composables/src/useNotifications.ts:96]: okay, so then just typing needs to change so we won't need to do casting, so this internal method needs to require `options.type`

Comment (mdanilowicz) [reply packages/composables/src/useNotifications.ts:96]: done

File: packages/composables/src/useOrderPayment.test.ts [hunk 11]
```diff
@@ -32,5 +32,49 @@ describe("useOrderPayment", () => {
...
+      expect.objectContaining({
+        body: {
+          orderId: "018c6832f1a1731db276839b3f251fc2",
+          paymentMethodId: "test",
+        },
+      }),
+    );
+  });
+
+  it("empty order", async () => {
```
[the remark this answers is not in the document]
Comment (mdanilowicz) [reply packages/composables/src/useOrderPayment.test.ts]: done

File: packages/composables/src/usePrice.test.ts [hunk 12]
```diff
@@ -1,12 +1,50 @@
...
+const currentNavigator = global.navigator;
+
+vi.mock("./useSessionContext.ts");
+const sessionContext = ref();
+
+vi.mocked(useSessionContext).mockReturnValue({
+  sessionContext,
+} as unknown as ReturnType<typeof useSessionContext>);
 
 describe("usePrice", () => {
```
[the remark this answers is not in the document]
Comment (mdanilowicz) [reply packages/composables/src/usePrice.test.ts:18]: done

**Review by patzick — APPROVED** [review]
